clang/test/SemaSYCL/intel-restrict.cpp

```diff
@@ -1,10 +1,18 @@
 // RUN: %clang_cc1 %s -fsyntax-only -fsycl-is-device -sycl-std=2017 -Wno-sycl-2017-compat -triple spir64 -DCHECKDIAG -verify
 // RUN: %clang_cc1 %s -fsyntax-only -ast-dump -fsycl-is-device -sycl-std=2017 -Wno-sycl-2017-compat -triple spir64 | FileCheck %s
 
-[[intel::kernel_args_restrict]] void func_do_not_ignore() {}
+// Test checks support and functionality of [[intel::kernel_args_restrict]] attribute.
+
+[[intel::kernel_args_restrict]] void func_do_not_ignore() {} // OK
+
+void func_ignore() [[intel::kernel_args_restrict]] {} // expected-warning{{unknown attribute 'kernel_args_restrict' ignored}}
```

Comment:
I don't understand why we need to accept this syntax at all. It sounds like this attribute is supposed to be a function attribute. I don't think is a legal position for function attributes and so the compiler (correctly) throws an error. I don't see the need to change this to a warning. Am I misunderstanding something here ?

Reply:
According to the spec: https://github.com/triSYCL/sycl/blob/sycl/unified/master/sycl/doc/extensions/KernelRestrictAll/SYCL_INTEL_kernel_restrict_all.asciidoc, the intel::kernel_args_restrict attribute has an effect when applied to a function, and no effect otherwise.

So We should not give an error here. The purpose of this patch to ignore the attribute with a warning for not legal position the attribute is applied to

Reply:
In that case, I think we should push back on the spec instead.  Having a spec require we not error on obviously invalid code is user hostile.  IMO, this should be an error anyway, and the spec can fix itself to match the behavior of the compiler.

Reply:
I think the example usage in the spec is incorrect as well - https://github.com/triSYCL/sycl/blob/sycl/unified/master/sycl/doc/extensions/KernelRestrictAll/SYCL_INTEL_kernel_restrict_all.asciidoc

That is not a valid position in `functor_foo` right?

Reply:
Yeah, this is an error. I don't support downgrading it into a warning.

Reply:
Yep, that is correct.  In fact, it is invalid in BOTH examples, but the attribute one we are allowing for suspicious reasons.  In reality, we should disable that spelling and just implement: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2173r1.pdf

Reply:
Thanks @erichkeane and @elizabethandrews for the reviews and feedbacks. 
```

class functor_foo {
void operator()() [[intel::kernel_args_restrict]] { }
};
```
Currently we give an error: 'kernel_args_restrict' attribute cannot be applied to types.

 i will close this PR as we decided to keep the error here.

Reply:
> Yep, that is correct. In fact, it is invalid in BOTH examples, but the attribute one we are allowing for suspicious reasons. In reality, we should disable that spelling and just implement: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2173r1.pdf

Good news, I already implemented that in Clang. Specifically for the SYCL folks. :-D

Reply:
Wonderful!  Then we should probably start emitting a 'deprecated' notice in SYCL for any attribute with "SupportsNonconformingLambdaSyntax" in that position so we can start forcing folks to remove them.

Reply:
> I think the example usage in the spec is incorrect as well - https://github.com/triSYCL/sycl/blob/sycl/unified/master/sycl/doc/extensions/KernelRestrictAll/SYCL_INTEL_kernel_restrict_all.asciidoc
> 
> That is not a valid position in `functor_foo` right?

Yes, i have asked spec authors  to fix the example in the spec.